sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsublite/internal/MemoryLimiterImpl.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.gcp.pubsublite.internal;
+
+import static org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Preconditions.checkState;
+
+import javax.annotation.concurrent.GuardedBy;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class MemoryLimiterImpl implements MemoryLimiter {
+  private static final Logger LOG = LoggerFactory.getLogger(MemoryLimiterImpl.class);
+  private final long minBlockSize;
+  private final long maxAvailable;
+
+  @GuardedBy("this")
+  private long available;
+
+  public MemoryLimiterImpl(long minBlockSize, long maxAvailable) {
+    this.minBlockSize = minBlockSize;
+    this.maxAvailable = available;
+    this.available = maxAvailable;
+  }
+
+  @Override
+  public synchronized Block claim(long toAcquire) {
```

Comment:
any concerns about fairness/prioritization?
Seems like you could have one highthroughput subscriber starve another subscriber with a lot of backlog.

Reply:
The MemoryLimiter logic should over time reduce the amount of memory allocated to subscribers if near or over the max limit.

For example, assume the server is out of memory (available == 0) and one subscriber holds the max amount (100 MiB). There are some amount of servers holding the min memory (4 MiB). When the subscriber holding the max releases its block, the limiter has 100 MiB available, but won't give out more than half unless it would be below the min. So it will give back only 50 MiB, and retain the other 50 for itself. Then assume one of the subscribers only allocated 4 MiB comes back and requests 100 (the max) again. It will return its 4 MiB, then request from the pool of 54 MiB available, which the limiter will then give it 26 MiB (half) instead, while having 26 MiB remaining. Over time, this will equalize all subscribers (assuming the backend is keeping up with them so they always request the max limit)